src/Microsoft.Diagnostics.Monitoring/DiagnosticServices.cs

```diff
@@ -0,0 +1,118 @@
+﻿using Microsoft.Diagnostics.Monitoring.RestServer;
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring
+{
+    public sealed class DiagnosticServices : IDiagnosticServices
+    {
+        private const int MaxTraceSeconds = 60 * 5;
+
+        private readonly ILogger<DiagnosticsMonitor> _logger;
+        private readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
+
+        public DiagnosticServices(ILogger<DiagnosticsMonitor> logger)
+        {
+            _logger = logger;
+        }
+
+        public IEnumerable<int> GetProcesses()
+        {
+            //TODO This won't work properly with multi-container scenarios that don't share the process space.
+            //TODO We will need to use DiagnosticsAgent if we are the server.
+            return DiagnosticsClient.GetPublishedProcesses();
+        }
+
+        public async Task<Stream> GetDump(int pid, DumpType mode)
+        {
+            string dumpFilePath = FormattableString.Invariant($@"{Path.GetTempPath()}{Path.DirectorySeparatorChar}{Guid.NewGuid()}_{pid}");
+            NETCore.Client.DumpType dumpType = MapDumpType(mode);
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // Get the process
+                Process process = Process.GetProcessById(pid);
+                await Dumper.CollectDumpAsync(process, dumpFilePath, dumpType);
+            }
+            else
+            {
+                await Task.Run(() =>
+                {
+                    var client = new DiagnosticsClient(pid);
```

Comment:
Instead of creating a new instance of DiagnosticsClient each time these methods are invoked, wouldn't it be better to keep a single instance of DiagnosticsClient inside DiagnosticServices and use that each time?

Reply:
I'm reluctant to do that now, because this logic will get more complicated once the DiagnosticAgent api is introduced and i have two modes of operation. DiagnosticsClient is very lightweight and does nothing on construction, so we're not really gaining much by caching it. 

---
In reply to: [403439623](https://github.com/dotnet/diagnostics/pull/979#discussion_r403439623) [](ancestors = 403439623)